smartsim/_core/launcher/dragon/dragonBackend.py

```diff
@@ -63,93 +92,225 @@ class DragonBackend:
     by threads spawned by it.
     """
 
-    def __init__(self) -> None:
-        self._proc_groups: t.Dict[str, t.Tuple[ProcessGroup, t.List[int]]] = {}
+    def __init__(self, pid: int) -> None:
+        self._pid = pid
+        self._group_infos: t.Dict[str, ProcessGroupInfo] = {}
         self._step_id_lock = RLock()
+        self._hostlist_lock = RLock()
         self._step_id = 0
+        # hosts available for execution
+        # dictionary maps hostname to step_id of
+        # step being executed on it
+        self._initialize_hosts()
+        self._queued_steps: "collections.OrderedDict[str, DragonRunRequest]" = (
+            collections.OrderedDict()
+        )
+        self._running_steps: t.List[str] = []
+        self._completed_steps: t.List[str] = []
+
+        num_hosts = len(self._hosts)
+        host_string = str(num_hosts) + (" hosts" if num_hosts > 1 else " host")
+        self._shutdown_requested = False
+        self._updates = 0
+        print(host_string + f" available for execution: {self._hosts}")
```

Comment:
why concat w/the f-string already there!

`print(f"{host_string} available for execution: {self._hosts}")`

Reply:
Where's a facepalm emoji when you need it.

Reply:
I re-request a face palm emoji here!